src/coreComponents/common/KernelLaunchSelectors.hpp

```diff
@@ -0,0 +1,167 @@
+{
+namespace internal
+{
+
+template< typename T, typename LAMBDA >
+void kernelLaunchSelectorThermalSwitch( T value, LAMBDA && lambda )
+{
+  static_assert( std::is_integral< T >::value, "kernelLaunchSelectorThermalSwitch: type should be integral" );
+
+  switch( value )
+  {
+    case 0:
+    { lambda( std::integral_constant< T, 0 >() ); return; }
+    case 1:
+    { lambda( std::integral_constant< T, 1 >() ); return; }
+
+    default:
+    { GEOS_ERROR( "Unsupported thermal state: " << value ); }
+  }
+}
+
+template< typename T, typename LAMBDA >
+void kernelLaunchSelectorCompThermSwitch( T value, bool const isThermal, LAMBDA && lambda )
+{
+  static_assert( std::is_integral< T >::value, "kernelLaunchSelectorCompSwitch: value type should be integral" );
+
+  if( isThermal )
+  {
+    switch( value )
+    {
+      case 1:
+      {
+        lambda( std::integral_constant< T, 1 >(), std::integral_constant< T, 1 >() ); return;
+      }
+      case 2:
+      { lambda( std::integral_constant< T, 2 >(), std::integral_constant< T, 1 >() ); return; }
+      case 3:
+      { lambda( std::integral_constant< T, 3 >(), std::integral_constant< T, 1 >() ); return; }
+      case 4:
+      { lambda( std::integral_constant< T, 4 >(), std::integral_constant< T, 1 >() ); return; }
+      case 5:
+      { lambda( std::integral_constant< T, 5 >(), std::integral_constant< T, 1 >()); return; }
+      default:
+      { GEOS_ERROR( "Unsupported number of components: " << value ); }
+    }
+  }
+  else
+  {
+    switch( value )
+    {
+      case 1:
+      {
+        lambda( std::integral_constant< T, 1 >(), std::integral_constant< T, 0 >() ); return;
+      }
+      case 2:
+      { lambda( std::integral_constant< T, 2 >(), std::integral_constant< T, 0 >() ); return; }
+      case 3:
+      { lambda( std::integral_constant< T, 3 >(), std::integral_constant< T, 0 >() ); return; }
+      case 4:
+      { lambda( std::integral_constant< T, 4 >(), std::integral_constant< T, 0 >() ); return; }
+      case 5:
+      { lambda( std::integral_constant< T, 5 >(), std::integral_constant< T, 0 >() ); return; }
+      default:
+      { GEOS_ERROR( "Unsupported number of components: " << value ); }
+    }
+  }
+}
+
+
+template< typename T, typename LAMBDA >
+void kernelLaunchSelectorCompPhaseSwitch( T value, T n_phase, LAMBDA && lambda )
+{
+  static_assert( std::is_integral< T >::value, "kernelLaunchSelectorCompSwitch: value type should be integral" );
+
+  if( n_phase == 1 )
+  {
+    switch( value )
+    {
+      case 1:
+      {
+        lambda( std::integral_constant< T, 1 >(), std::integral_constant< T, 1 >() ); return;
+      }
+      case 2:
+      { lambda( std::integral_constant< T, 2 >(), std::integral_constant< T, 1 >() ); return; }
+      case 3:
+      { lambda( std::integral_constant< T, 3 >(), std::integral_constant< T, 1 >() ); return; }
+      case 4:
+      { lambda( std::integral_constant< T, 4 >(), std::integral_constant< T, 1 >() ); return; }
+      case 5:
```

Comment:
and I would do something similar here too.

Reply:
````c++
auto invokeLambda = [&] (auto numComp, T numPhases) {
    if (numPhases == 1) {
        lambda(numComp, std::integral_constant<T, 1>{});
    } else if (numPhases == 2) {
        lambda(numComp, std::integral_constant<T, 2>{});
    }
    else if (numPhases == 3)
    {
    lambda(numComp, std::integral_constant<T, 3>{});
    }
    else
    { /// error}
};
````

Reply:
done